Add Geocoding to the Client

boldtrn opened the pull request:
Hi,

so I just started to add GeoCoding to the client. Right now I am not sure how to proceed.

I added the GHGEntry and GHGResponse, which are basically copied from the [graphhopper-geocoder-converter](https://github.com/boldtrn/graphhopper-geocoder-converter/tree/master/src/main/java/com/graphhopper/converter/api). In the converter we use Jackson to convert Json to JavaObjects, but in the client right now we don't use jackson.

So right now I am thinking if I should add Jackson as dependency or if I should do the parsing by hand? 

Adding dependencies is not really nice, but on the other hand, this would make the code more maintainable and we could maybe add the geocoder-converter as dependency and don't need to copy the classes? 

What do you think is the better way to do this?

BTW: Code quality is pretty bad right now, I just created the PR to discuss the various options.

karussell commented on client/src/main/java/com/graphhopper/api/GraphHopperGeocoder.java:
```diff
@@ -0,0 +1,73 @@
+                build();
+    }
+
+    public GHGResponse geocode(String search, String locale) {
```
Here I would introduce a GHGRequest to avoid expanding the parameter list later further and breaking the API again and again

boldtrn replied:
Makes sense, thanks :+1:

karussell commented:
Thanks, this looks good. Although I'm not sure about the naming. Why not just GeocodeEntry and GeocodeResponse?

Regarding the library: we currently use json.org https://github.com/graphhopper/directions-api-java-client/blob/master/client/pom.xml#L51

Regarding the converter&dependency: that is indeed a good idea, but then we should probably create a new 'geocoding' module making it separate from the routing?

## Discussion the future :)

Another idea would be to define a swagger.json like we already do for route optimization and have the clients created automatically (in several more languages), this is also not that complex but I did not choose this direction for the routing as I wanted the same API for desktop, mobile and as a client. But using the swagger approach has these advantages. (One thing one would need: if polyline encoding is enabled one needs a method which must be handcoded, but this is just a performance/bandwidth tuning we could do in all languages at a later point)

devemux86 commented:
>Regarding the converter&dependency: that is indeed a good idea, but then we should probably create a new 'geocoding' module making it separate from the routing?

:+1: 

I'd suggest a common JSON dependency for all GraphHopper modules, as many times they're used together.
Also not forget that in Android it's best to not overdo it with too many dependencies.

devemux86 commented on client/src/main/java/com/graphhopper/api/GraphHopperGeocoder.java:
```diff
@@ -0,0 +1,73 @@
+            }
+
+            if (locale != null || !locale.isEmpty()) {
+                url += "&key=" + WebHelper.encodeURL(locale);
```
`key` here is wrong parameter?

boldtrn replied:
Thanks, yes, as I said the quality is really awful right now and definitely not tested

boldtrn commented:
> Why not just GeocodeEntry and GeocodeResponse?

I tried to stay close to existing conventions with the GHMRequest and GHMResponse. I am fine with renaming them accordingly?

> but then we should probably create a new 'geocoding' module making it separate from the routing?

I also thought about that, but the GeoCoding belongs to the directions API, so I thought it should live in that module.

> Another idea would be to define a swagger.json

How about doing that now? I mean we could only create a swagger.json for the GeoCoding in the beginning. If I understand the concept correct this would parse the GeoCoding Json for us, but we need to define the format in a swagger.json, right?

> I'd suggest a common JSON dependency for all GraphHopper modules, as many times they're used together.

+1, different dependencies only create problems.

karussell commented:
> I tried to stay close to existing conventions with the GHMRequest and GHMResponse. I am fine with renaming them accordingly?

Oh, you are right. GHG reads a bit ugly, but then we should keep it now.

> I also thought about that, but the GeoCoding belongs to the directions API, so I thought it should live in that module.

What do you mean here? We already have an route-optimization module and could add a 'geocoding' and rename the current 'client' to 'routing'?

> How about doing that now? I mean we could only create a swagger.json for the GeoCoding in the beginning. 

Yes, might be a start.

> If I understand the concept correct this would parse the GeoCoding Json for us, but we need to define the format in a swagger.json, right?

We would need to define input and output but get clients in different languages (and documentation) for free

boldtrn commented:
> What do you mean here? We already have an route-optimization module and could add a 'geocoding' and rename the current 'client' to 'routing'?

Ok :)

> We would need to define input and output but get clients in different languages (and documentation) for free

Ok I will try creating a swagger.json, I'll keep you updated :)

boldtrn commented:
> Ok I will try creating a swagger.json, I'll keep you updated :)

Ok, so I created a simple swagger.json (actually a swagger.yaml) and created a client.

Right now, there are some things I am worried about:
- The code is untested (as well as the optimization module) (I created some requests by hand)
 - I would prefer to write some tests, the test file is not overwritten by the code generator, AFAIK. What do you think?
-  I handle errors different than the optimization module (not sure if this is important)
- I have not split the reverse geocoding from the forward geocoding, as it is the same endpoint
- We might also want to move the swagger.yaml to the graphhopper server
- Probably a lot more ;)

karussell commented:
> Ok, so I created a simple swagger.json 

Nice - thanks :) !

> (actually a swagger.yaml) 

Is this simpler or the new default?

> as well as the optimization module

Oh, indeed, ugly. We have internal tests for this but for now preferred to keep this close to the API and avoid duplication, hmmh probably better here and reuse the client test code in our server backend.

> I would prefer to write some tests, the test file is not overwritten by the code generator, AFAIK. What do you think?

Yes, we should create some, also we can then use them also as examples. (+ we could use the examples in the route opt module as tests)

> I handle errors different than the optimization module (not sure if this is important)

What do you mean here?

> I have not split the reverse geocoding from the forward geocoding, as it is the same endpoint

:+1: 

> We might also want to move the swagger.yaml to the graphhopper server

Yeah, this is another thing we have already duplicated in our website repo and backend. Will think about how to handle this.

> I have not added to many descriptions to the swagger.yaml yet, we might want to add more information here

Sure

karussell commented on geocoding/LICENSE:
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
We should move this to the parent... ah this is auto created (?)

boldtrn replied:
Yes, this is auto created

boldtrn replied:
Anything in the geocoding folder is auto generated

karussell commented on README.md:
````diff
@@ -139,3 +139,11 @@ The route optimization client is in the maven central repository, thus you can j
+## Generate from Swagger
+
+```
+java -jar ../swagger-codegen-cli.jar generate -i geocoding_swagger.yaml -l java -o geocoding -c geocoding-config.json
````
We have to streamline this somehow with the route opt client (but can do this later on). E.g. using the same [config stuff](https://github.com/graphhopper/directions-api-clients-route-optimization/blob/master/create.sh#L24-L25) e.g. it overwrite model and package names.

boldtrn commented:
> Is this simpler or the new default?

The Swagger Editor started with a yaml file, so I used that and changed it accordingly. Personally, I also think that yaml is better suited for config files than json, because it is easier to read and write. 

> here and reuse the client test code in our server backend.

I would be fine with a bunch of tests, that check if the client is producing correct results. For example a test case that uses every parameter available in one or two different settings to be sure that all parameters are used correctly. I am not sure if the client should test the backend itself?

> What do you mean here?

I have not looked at the route-optimization swagger when creating the swagger for the geocoding. Afterwards I realized, that in the route-opt swagger errors are handled like that:

```
400: {
description: "Error occurred when reading client request. Request is invalid."
},
500: {
description: "Error occurred on server side."
}
``` 

But I wrote:
```
        200:
          description: An array found locations
          schema:
            $ref: '#/definitions/GeocodingResult'
        default:
          description: Unexpected error
          schema:
            $ref: '#/definitions/Error'

-----------------------------------------

  Error:
    type: object
    properties:
      message:
        type: string
```

This is parsing the error message and returns it. This was more or less given in the sample API of the swagger editor, so I just reused that.

boldtrn commented:
Ok, so I added a test plan for the geocoding module. I stumbled across several things, which I marked as TODO. @karussell maybe you could add some comments to the todos?

Besides the Todos, it seems that the Documentation for the Geocoding API is outdated? The example result does not match the actually returned result. We added several new fields, for example osm_type, also many more.

boldtrn commented on geocoding_swagger.yaml:
```diff
@@ -0,0 +1,123 @@
+          required: false
+          type: integer
+          format: int32
+        - name: debug
```
Do we want to add debug to the client?

karussell replied:
As the outcome is not really accessible to the client I would remove it?

boldtrn commented:
This branch is ready for review, I am finished with my changes :).

karussell commented on geocoding/src/test/java/io/swagger/client/api/GeocodingApiTest.java:
```diff
@@ -0,0 +1,175 @@
+        GeocodingResult response = api.geocodeGet(key, q, locale, limit, debug, reverse, point, provider);
+        // Default Limit and Locale
+        // TODO: Why is default limit 15? Documentation says it's 10
+        // see: https://graphhopper.com/api/1/docs/geocoding/
```
Indeed the current default is 15

karussell commented on geocoding/src/test/java/io/swagger/client/api/GeocodingApiTest.java:
```diff
@@ -0,0 +1,175 @@
+        GeocodingResult response = api.geocodeGet(key, q, locale, limit, debug, reverse, point, provider);
+
+        // Default Limit and Locale
+        // TODO: Why is it 5 in this case?
```
Indeed this difference is not optimal

karussell commented on geocoding/src/test/java/io/swagger/client/api/GeocodingApiTest.java:
```diff
@@ -0,0 +1,175 @@
+
+        // Default Limit and Locale
+        assertEquals(1,response.getHits().size());
+        // TODO Why is locale null here? Maybe we should add it to the geocoding-converter?
```
Indeed. Thanks for catching all these issues!

karussell commented:
Merged and listed the other issues here https://github.com/graphhopper/directions-api/issues/39

karussell commented:
One tiny thing: maybe we need to point to the tests in the readme as an example?
